### ISSUE TITLE
Why the fine-tuned CNN just has one branch

### ISSUE THREAD

**Issue by iiwindii** [filipradenovic/cnnimageretrieval#1]
The finetuned CNN is a Siamese network which usually has two branches. However, I found  each of the fine-tuned models you provided has one branch. Is this because the training samples are tuples consisting of query image, positive image and several negative images instead of images? And also, when you compute the contrastive loss, it is computed on the tuple which actually consists of 6 image pairs？ The batchsize is set to 5, does that mean one batch consists of 5 tuples and then 30 image pairs in total?
Look forward to your reply, thanks in advance!

**Comment by filipradenovic** [comment]
>  I found each of the fine-tuned models you provided has one branch. Is this because the training samples are tuples consisting of query image, positive image and several negative images instead of images? 

At training time we process several tuples at a time (batchSize=5 represents 5 tuples at a time). One tuple consists of 1 query (anchor), 1 positive and numNegative negative images (numNegative=5 in our experiments), which constitutes 6 pairs for contrastive loss (1 query-positive pair and 5 query-negative pairs) and 5 triples for triplet (ranking) loss. Both contrastive loss and triplet loss are computed over a tuple, so smallest input during training (at this moment) is one tuple (batchSize=1). Having one branch is beneficial because of: (i) smaller memory consumption (we avoid keeping 2 or 3 identical branches in the memory); (ii) same architecture can be reused for both contrastive (2 branches) and triplet (3 branches) loss computed over a tuple; and (iii) any other loss computed over a tuple can be easily plugged in.

> When you compute the contrastive loss, it is computed on the tuple which actually consists of 6 image pairs?

Yes, whole tuple is processed in the same batch. Therefore, we feed 7 images to the network that represent 6 pairs. In a naive approach, when the query image is different for each pair, 6 pairs require 12 images.

>  The batchsize is set to 5, does that mean one batch consists of 5 tuples and then 30 image pairs in total?

Yes, 30 image pairs in the case when we use 5 tuples and 5 negative images per tuple (batchSize=5 & numNegative=5).

Finally, at test time (only forward pass computed, with L2 normalization layer used as a final layer), one image is processed at a time, ie, an image is embedded in a D-dimensional vector space. So multiple-branch architecture is not necessary.

I hope this resolves all your issues. Any additional question is always welcome.

**Comment by iiwindii** [comment]
@filipradenovic Thanks for your detailed response. I can understand clearly now. Having one branch is really an awesome idea to save memory. I still have two more questions: (1) have you ever tried having two branches for contrastive loss or three branches for triplet loss in your experiments?  Though the branches share the same weights, the performance should be different (maybe just a bit) from having one branch. (2) I want to adapt your code to implement one-branch Siamese network and one-branch Triplet network. But the training samples are image pairs (two images) for Siamese network and triples (three images) for Triplet network. Could you provide some advice?

**Comment by filipradenovic** [comment]
> (1) have you ever tried having two branches for contrastive loss or three branches for triplet loss in your experiments? Though the branches share the same weights, the performance should be different (maybe just a bit) from having one branch.

We have tried, and these two scenarios are identical. Both losses are computed using pairwise distances which will be identical with one-, two-, or three-branch implementation. As I have said before, one branch is implemented only due to convenience and lower memory requirements per batch.

> (2) I want to adapt your code to implement one-branch Siamese network and one-branch Triplet network. But the training samples are image pairs (two images) for Siamese network and triples (three images) for Triplet network. Could you provide some advice?

For triplet network you can easily set numNegative=1 and you will have a triplet instead of a tuple. For contrastive if you set numNegative=1 you will have two pairs instead of a tuple (1 pos and 1 neg pair with same anchor image). If you want pairs exclusively you will have to make small implementation changes in get_batch.m to pack pairs instead of tuples. As an example, if you want to pack pos-neg-pos-neg-neg pairs you will have pack pairs of images in input_ims; and you will have input_lbls = [-1 1 -1 0 -1 1 -1 0 -1 0].

**Comment by iiwindii** [comment]
Thank you for your advice! I will try that!

**Comment by rizwanasif** [comment]
@fengyunxiaozi can i have your email please? i want to discuss some issues i am facing while training. One, I already discussed with you.. i have some more questions... qq or wechat, please?
